### ISSUE TITLE
Subword motion ignores line breaks, jumps to next line unexpectedly

### ISSUE THREAD

**Issue by ribelo** [zed-industries/zed#17780]
## Description
The Subword motion feature is not working as expected when it encounters end-of-line characters. Instead of treating the end of the line as a boundary, it immediately jumps to the subword end of the next word on the line below.

## Steps to Reproduce
1. Open a file in Zed editor
2. Place the cursor at the beginning of the last subword in a line
3. Use the Subword motion command to move forward

## Expected Behavior
The cursor should move to the end of the current word and stop at the end of the line.

## Actual Behavior
The cursor skips the end-of-line character(s) and jumps to the subword end of the first word on the next line.

### Environment

Zed: v0.152.3 (Zed)
OS: Linux Wayland nixos 24.11
Memory: 30.6 GiB
Architecture: x86_64
GPU: AMD Radeon Graphics (RADV REMBRANDT) || radv || Mesa 24.2.2

### If applicable, add mockups / screenshots to help explain present your vision of the feature

<details><summary>video</summary>

https://github.com/user-attachments/assets/b8d3741b-2c2e-456e-85c2-ad306981bb4b

</details>

<details><summary>screenshots</summary>

![2024-09-13-092026_grim](https://github.com/user-attachments/assets/275a606b-2dfe-482a-bf8e-957254727078)
![2024-09-13-092037_grim](https://github.com/user-attachments/assets/dac985ee-4efc-4f1c-b3e7-db9dcf806480)
![2024-09-13-092044_grim](https://github.com/user-attachments/assets/16160c2b-704b-4384-b5b4-965096d98ee1)

</details>

**Comment by 0x2CA** [comment]
This is normal behavior, because it uses multiple lines to search, which simulates the movement of a word.

```rust
fn next_subword_start(
    map: &DisplaySnapshot,
    mut point: DisplayPoint,
    ignore_punctuation: bool,
    times: usize,
) -> DisplayPoint {
    let classifier = map
        .buffer_snapshot
        .char_classifier_at(point.to_point(map))
        .ignore_punctuation(ignore_punctuation);
    for _ in 0..times {
        let mut crossed_newline = false;
        let new_point = movement::find_boundary(map, point, FindRange::MultiLine, |left, right| {
            let left_kind = classifier.kind(left);
            let right_kind = classifier.kind(right);
            let at_newline = right == '\n';

            let is_word_start = (left_kind != right_kind) && !left.is_alphanumeric();
            let is_subword_start =
                left == '_' && right != '_' || left.is_lowercase() && right.is_uppercase();

            let found = (!right.is_whitespace() && (is_word_start || is_subword_start))
                || at_newline && crossed_newline
                || at_newline && left == '\n'; // Prevents skipping repeated empty lines

            crossed_newline |= at_newline;
            found
        });
        if point == new_point {
            break;
        }
        point = new_point;
    }
    point
}
```

**Comment by lionel-** [comment]
hmm it makes subword motions pretty much unusable for me and is inconsistent with non-subword motions. I have a hard time seeing how this could be expected behaviour?

In this screencast I type `e` twice:

https://github.com/user-attachments/assets/d707df91-2484-404c-93c2-ba5a6b28ad23

**Comment by 0x2CA** [comment]
> hmm it makes subword motions pretty much unusable for me and is inconsistent with non-subword motions. I have a hard time seeing how this could be expected behaviour?
> 
> In this screencast I type `e` twice:
> 
>  Screen.Recording.2025-01-06.at.11.08.15.mov

Oh, sorry I misunderstood, indeed there is a problem.

When using e, it will unexpectedly skip many words and may even skip many lines.

The key code is here

```rust
pub(crate) fn next_subword_end(
    map: &DisplaySnapshot,
    mut point: DisplayPoint,
    ignore_punctuation: bool,
    times: usize,
    allow_cross_newline: bool,
) -> DisplayPoint {
    let classifier = map
        .buffer_snapshot
        .char_classifier_at(point.to_point(map))
        .ignore_punctuation(ignore_punctuation);
    for _ in 0..times {
        let new_point = next_char(map, point, allow_cross_newline);

        let mut crossed_newline = false;
        let mut need_backtrack = false;
        let new_point =
            movement::find_boundary(map, new_point, FindRange::MultiLine, |left, right| {
                let left_kind = classifier.kind(left);
                let right_kind = classifier.kind(right);
                let at_newline = right == '\n';

                if !allow_cross_newline && at_newline {
                    return true;
                }

                let is_word_end = (left_kind != right_kind) && !right.is_alphanumeric();
                let is_subword_end =
                    left != '_' && right == '_' || left.is_lowercase() && right.is_uppercase();

                let found = !left.is_whitespace() && !at_newline && (is_word_end || is_subword_end);

                if found && (is_word_end || is_subword_end) {
                    need_backtrack = true;
                }

                crossed_newline |= at_newline;
                found
            });
        let mut new_point = map.clip_point(new_point, Bias::Left);
        if need_backtrack {
            *new_point.column_mut() -= 1;
        }
        if point == new_point {
            break;
        }
        point = new_point;
    }
    point
}
```

If you replace `if !allow_cross_newline && at_newline { return true; }` with `if at_newline { return true; }` here, it will behave normally.

I don't quite understand the function of the `allow_cross_newline` parameter.

**Comment by lionel-** [comment]
I see the expected behaviour when the line ends with punctuation. When it doesn't, the last subword on the line is skipped. Here is a video where I type `e` exactly twice. On the second time the cursor jumps too far:

https://github.com/user-attachments/assets/ac2f8a34-8d13-4569-9fe3-39dccdfcee61

**Comment by oluckyman** [comment]
Still relevant. I have the same issue. 
Can reproduce on Zed: v0.191.5 (Zed) 
OS: macOS 14.7.4
Memory: 16 GiB
Architecture: aarch64

**Comment by sylbru** [comment]
In case it’s useful, this bug on `vim: next subword end` for the last word of a line is still happening. It’s still very annoying on a daily basis for those of us that use subword motions. Sorry I don’t have the knowledge to look into this myself. 

Zed: v0.211.6 (Zed) 
OS: Linux X11 linuxmint 22.1
Memory: 30.6 GiB
Architecture: x86_64
GPU: AMD Radeon Graphics (RADV RENOIR) || radv || Mesa 25.0.7-0ubuntu0.24.04.2

**Comment by lionel-** [comment]
yep very annoying, and my vim motion patterns have already started to change in weird ways to adjust for this. If no one has fixed it yet I might take a morning during the christmas break to squash this bug before it breaks my vim brain too much 😅

**Comment by ConradIrwin** [comment]
Please do! Happy to pair if helpful: https://cal.com/conradirwin/pairing
